### PR TITLE
[numpy] Shape support scalar tensor

### PR DESCRIPTION
**Pull request by reminisce**
## Description ##
This PR made necessary changes in the shape-related infra code to support scalars, aka, zero-dim tensors. 
See this [RFC](https://github.com/apache/incubator-mxnet/issues/14253) for more information.

Implemented `np.sum` to verify the correctness with the following script.

After merging this PR, we can get started to make all the existing infer shape functions accommodate the new definition of empty shapes. Implementation of NumPy operators can be carried out in parallel.

Note that before we change all the existing infer functions to accommodate the new empty shape definition, the CI will fail without surprise. When we work on specific operators, we can test those locally to verify the correctness and submit PRs. After all the infer functions are revised, the CI should pass eventually. Then, we can merge the dev branch to the master.

```python
import mxnet as mx
from mxnet import numpy as np


shape = (2, 2)
ret = np.sum(mx.nd.ones(shape))
print("Imperative invoke result==============")
print(ret)

data = mx.sym.var('data', shape=shape)
ret = mx.sym.numpy.sum(data)
exe = ret.simple_bind(ctx=mx.cpu(), data=(2, 2))
exe.forward(data=mx.nd.ones(shape))
print("Symbol bind result===============")
print(exe.outputs[0])

print("CachedOp result===============")
func = mx.nd.CachedOp(ret)
print(func(mx.nd.ones(shape)))
```
```
Imperative invoke result==============

4.0
<NDArray  @cpu(0)>
Symbol bind result===============

4.0
<NDArray  @cpu(0)>
CachedOp result===============

4.0
<NDArray  @cpu(0)>
```
@junrushao1994 @szha @eric-haibin-lin @zheng-da @yzhliu

### CODE REVIEW

**Comment by junrushao1994** [comment]
Awesome work! Let’s work on this

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-awaiting-review]

File: include/mxnet/tuple.h [hunk 1]
```diff
@@ -106,7 +113,10 @@ class Tuple {
   inline void assign(RandomAccessIterator begin,
                      RandomAccessIterator end) {
     this->SetDim(end - begin);
-    std::copy(begin, end, this->begin());
+    CHECK_GE(ndim(), 0);
+    if (ndim() > 0) {
```
Comment (junrushao1994) [inline include/mxnet/tuple.h]: Don't need check `ndim() > 0` actually

File: include/mxnet/tuple.h [hunk 2]
```diff
@@ -186,6 +201,7 @@ class Tuple {
    * \return the corresponding dimension size
    */
   inline ValueType& operator[](size_t i) {
```
Comment (junrushao1994) [inline include/mxnet/tuple.h]: Let's make `int i` instead

File: include/mxnet/tuple.h [hunk 3]
```diff
@@ -186,6 +201,7 @@ class Tuple {
...
+    CHECK_GT(ndim(), 0);
```
Comment (junrushao1994) [inline include/mxnet/tuple.h]: Don't check this, but let's check 0 <= i < ndim

File: include/mxnet/tuple.h [hunk 4]
```diff
@@ -194,6 +210,7 @@ class Tuple {
    * \return the corresponding dimension size
    */
   inline const ValueType& operator[](size_t i) const {
+    CHECK_GT(ndim(), 0);
```
Comment (junrushao1994) [inline include/mxnet/tuple.h]: Same as above

File: include/mxnet/tuple.h [hunk 5]
```diff
@@ -220,6 +237,7 @@ class Tuple {
    * \return the ostream
    */
   friend std::ostream &operator<<(std::ostream &os, const Tuple<ValueType> &t) {
+    if (t.ndim() == -1) return os;
```
Comment (junrushao1994) [inline include/mxnet/tuple.h]: Why not we print something like "?"?

File: src/operator/operator_common.h [hunk 6]
```diff
@@ -159,16 +169,16 @@ inline std::string type_string(const int& x) {
  * \return whether x and y are compatible.
  */
 inline bool shape_assign(mxnet::TShape *y, const mxnet::TShape& x) {
-  if (y->ndim() == 0) {
+  if (y->ndim() < 0) {
```
Comment (junrushao1994) [inline src/operator/operator_common.h]: Let's say `== -1` instead of `< 0`

File: include/mxnet/tensor_blob.h [hunk 7]
```diff
@@ -196,6 +196,10 @@ class TBlob {
     CHECK(mshadow::DataType<DType>::kFlag == type_flag_)
       << "TBlob.get_with_shape: data type do not match specified type."
       << "Expected: " << type_flag_ << " v.s. given " << mshadow::DataType<DType>::kFlag;
+    // if scalar, handle it with special care
+    if (ndim() == 0) {
+      return this->get_with_shape<Device, 2, DType>(mshadow::Shape2(1, 1), stream);
```
Comment (szha) [inline include/mxnet/tensor_blob.h]: why is this needed? should the `shape_.FlatTo2D()` handle it, given that the next call already passes `dptr_`?

Comment (reminisce) [reply include/mxnet/tensor_blob.h]: Because this line would crash.
```cpp
shape_[shape_.ndim() - 1],
```

File: include/mxnet/tuple.h (hunk 1 above)
Comment (reminisce) [reply include/mxnet/tuple.h]: Good catch. I was fixing another assignment problem and just added the check everywhere. Will remove this.

File: include/mxnet/tuple.h (hunk 5 above)
Comment (reminisce) [reply include/mxnet/tuple.h]: How about making it more specific: `UNKNOWN_SHAPE`?

File: include/mxnet/tensor_blob.h (hunk 7 above)
Comment (szha) [reply include/mxnet/tensor_blob.h]: Is the stride needed?

Comment (szha) [reply include/mxnet/tensor_blob.h]: Without providing the stride, it already looks up the size of this dimension as stride:
`static const int  kSubdim = dimension - 1;`

File: include/mxnet/tuple.h [hunk 8]
```diff
@@ -404,9 +452,11 @@ class TShape : public Tuple<dim_t> {
   }
   /*! \return total number of elements in the shape */
   inline size_t Size() const {
+    if (this->ndim() == -1) return 0;
```
Comment (szha) [inline include/mxnet/tuple.h]: Should this throw an error instead?

Comment (szha) [reply include/mxnet/tuple.h]: The hope is to easily identify any illegal use of undefined shapes. what do you think?

Comment (reminisce) [reply include/mxnet/tuple.h]: Agree. It doesn't make sense to get a size value when shape is unknown. Will change that.

File: include/mxnet/tuple.h [hunk 9]
```diff
@@ -220,6 +235,10 @@ class Tuple {
    * \return the ostream
    */
   friend std::ostream &operator<<(std::ostream &os, const Tuple<ValueType> &t) {
+    if (t.ndim() == -1) {
```
Comment (szha) [inline include/mxnet/tuple.h]: should there be an explicit method for checking whether shape is known?

Comment (reminisce) [reply include/mxnet/tuple.h]: By the current convention, it's defined in `operator_common.h` as `shape_is_known`, previously known as `shape_is_none`. For this case specifically, we should not use that function, because we want to just check `ndim`. For this release, I suggest we not move it to keep the convention as is. For the next major release, we can move the util function closer to where the object is defined.

**Comment by szha** [comment]
Thanks for addressing my concerns

File: include/mxnet/tuple.h [hunk 10]
```diff
@@ -316,48 +335,74 @@ class Tuple {
...
   /*!
    * constructor to construct a shape with all 1.
+   * TODO(junwu): The value should default to -1. Need to keep 1 for now
+   * for backward compatibility. Change it to -1 in the future when we can
+   * break backward compatibility.
    * \param ndim the number of dimension
+   * \param value the dimension size for all dims
    */
-  inline TShape(uint32_t ndim) {  // NOLINT(*)
+  inline TShape(int ndim, int value = 1) {  // NOLINT(*)
```
Comment (junrushao1994) [inline include/mxnet/tuple.h]: Should we make `value = -1` now, or in the future? I mean, this PR is targeted at the dev branch, so should be good anyway

File: include/mxnet/tuple.h [hunk 11]
```diff
@@ -404,9 +453,11 @@ class TShape : public Tuple<dim_t> {
   }
   /*! \return total number of elements in the shape */
   inline size_t Size() const {
+    CHECK_GE(this->ndim(), 0) << "Shape is unknown.";
```
Comment (junrushao1994) [inline include/mxnet/tuple.h]: Is this 0 or -1?

File: include/mxnet/tuple.h (hunk 10 above)
Comment (reminisce) [reply include/mxnet/tuple.h]: I made it -1 before, but it failed at other places where 1 is expected. To avoid resolving conflicts with the master branch too early, we can defer this to the time when we want to merge the whole changes back to master.

Comment (junrushao1994) [reply include/mxnet/tuple.h]: Sounds good

File: include/mxnet/tuple.h (hunk 11 above)
Comment (reminisce) [reply include/mxnet/tuple.h]: It's CHECK_GE, so should be 0. :)

Comment (junrushao1994) [reply include/mxnet/tuple.h]: I see. I misread just now. Suggest using `CHECK_NE` to be more specific

Comment (reminisce) [reply include/mxnet/tuple.h]: I think `>= 0` is a safer check than `!= -1` to rule out the cases when ndim is mistakenly messed up ending with a negative integer other than -1.

Comment (junrushao1994) [reply include/mxnet/tuple.h]: When it is a negative number other than -1, you should immediately throw an error, instead of silently allow this to happen.

Comment (reminisce) [reply include/mxnet/tuple.h]: Makes sense. I will add this check wherever possible.
